self-paced-labs/image_build/kaniko_lab/README.md

```diff
@@ -0,0 +1,200 @@
+# Kaniko 教學
+Kaniko 是 Google 在 2018 年釋出的開源容器映像檔工具，最大的特點就是不需依賴於 Docker Daemon，所有操作都是在 user-space 執行。
+
+## 前言
+通常在 CI/CD 中 build image 有下列幾種方式：
+### Use Docker-in-Docker
+在容器中執行 `docker` 指令，會在 privileged mode 下執行 build image 相關操作。
+但必須在 privileged mode 底下執行，以及無法使用 layer caching 這類的快取機制，官方也較不推薦。
+
+### Use Docker socket binding 
+又被稱為 DooD (Docker-outside-of-Docker)，掛載 Host 端的 docker socket 到 container，讓 container 不需 Docker daemon 也能運行相關操作。
+但在掛載 `/var/run/docker.sock` 後，相當於環境沒有被完全隔離，可能與其他容器衝突。
+
+
+### Use other tool without privileged mode
+使用如本篇介紹的 Kaniko 或是 Buildah，主要是為了解決 Dind 的問題：
+- Docker-in-Docker 需要特權模式才能運行
+- Docker-in-Docker 通常會導致性能損失並且速度可能非常慢
+
+## 運作原理
+- 以下操作皆在 user-space 使用 executor image 執行，避免使用privileged mode，因此也不會需要 docker daemon。
+  - Kaniko 會從 Base image 提取 file system
+  - 當逐行執行 Dockerfile 時，在 user-space 對 filesystem 做快照，運用 cache 去和 base image 比對
+  - 將有更改的 files append 到 base image，並更新 image 的 metadata
+  - 當最後執行完整個 Dockerfile 時， Kaniko executor 將 build 完成的 image push 到指定的 registry 中。
+
+- 流程如下圖：
+- 根據顏色分為不同層級
+    * 紫色的是最外層的邏輯
+    * 藍色的是每個 stage
+    * 綠色的是執行 Dockerfile 裡面的 instruction
+
+![kaniko_workflow](image/kaniko_workflow.jpg)
```

Comment:
Please use the relative URL to frame the image.
`![kaniko_workflow](../image/kaniko_workflow.jpg)`

<img width="334" alt="CleanShot 2022-06-20 at 23 40 51@2x" src="https://user-images.githubusercontent.com/43361940/174637422-462a140e-d997-4c5c-922f-f48debe5712a.png">

Reply:
ok, fixed.